### DISCUSSION TITLE
Authelia Integration

### DISCUSSION THREAD

**Discussion by viharm** [firefly-iii/firefly-iii#6252] [Firefly III] [answered]
Since LDAP support was dropped in 5.7.0, I have been trying to migrate to _Authelia_. After weeks of trial and error, I've finally succeeded in integrating _Authelia_ into _FFIII_.

The guidance in the documentation is spot on, thanks. Especially the one about [empty guard header](https://docs.firefly-iii.org/firefly-iii/advanced-installation/authentication/#the-guard-header-was-unexpectedly-empty), where I had to specifically specify `AUTHENTICATION_GUARD_HEADER`=`HTTP_REMOTE_USER`

Since I am migrating from an instance which was using native LDAP, the old setup was such that the _FFIII_ username was the email address taken from the LDAP directory.

With the new auth setup using `remote_user_guard`, the [ _OpenLDAP_ → _Authelia_ → _NGINX_ ] chains sends the `uid` of the user in the `Remote-User` header, and not the email address.

Q1: Does this mean _FFIII_ will treat it as a new user? (old: `username@email.address`; new: `username`)

I have tried utilize the guidance on [customizing headers](https://docs.firefly-iii.org/firefly-iii/advanced-installation/authentication/#customizing-remote-user-header) and specified `AUTHENTICATION_GUARD_EMAIL`=`HTTP_REMOTE_EMAIL`, but the _FFIII_ username still appears to be the `uid` (I tried with a new user, just to be sure). Also, I check the `users` table in the database, and couldn't find the email anywhere.

Q2: Is the `AUTHENTICATION_GUARD_EMAIL` only for communicating with the user? Is it not store anywhere?

Q3: Can use the email from the header as the username? Like so... `AUTHENTICATION_GUARD_HEADER`=`HTTP_REMOTE_EMAIL`

Obviously my worry is to retain my existing data with months of accounts.

Thanks.

**Comment by JC5** [comment]
Hey,

Glad to hear it works as expected. I'll make sure your questions make it to the FAQ as well.

- Q1: Yes. You can safely rename the user in the database if you want to, but you may be able to setup Authelia in such a way that it puts the email address in the REMOTE_USER header, and not the username. I believe this is possible, but I'm not sure how.
- Q2: Yes. It is stored in the "preferences" table.
- Q3: Yes, as described in Q1.

Hope this helps, let me know if it works. Make sure you back up your database.

**Reply by viharm** [reply]
Thanks. It is actually very easy to tune _Authelia_ to send the email address in the `Remote-User` header. However, since I've now learned how to deploy _Authelia_ (thanks :-) ), I will be integrating other apps, for which I will retain defaults for `Remote-User`. So the decision is between...

1. `AUTHENTICATION_GUARD_HEADER`=`HTTP_REMOTE_USER` combined with `preferences`.`remote_guard_alt_email` (edit `users`.`email` in DB to username only)
1. `AUTHENTICATION_GUARD_HEADER`=`HTTP_REMOTE_EMAIL`

From a design _FFIII_ point-of-view is there a preference?

**Reply by JC5** [reply]
Either works, it doesn't matter from Firefly III's perspective.

**Reply by viharm** [reply]
Fantastic! Thanks. I've successfully migrated to _Authelia_ on my production instance, using my master DB (backed up first, obviously!). I'm now running 5.7.10 as a _Docker_ container. I'm encountering a curious situation where I'm getting an email `Caught an error in Firefly III`. The summary information is `The guard header was unexpectedly empty.`. What's really curious is that I don't encounter any error whilst using the instance. I noticed that the error email helpfully includes the headers it encountered as follows:
 - `host: localhost:8080`
- `user-agent: curl/7.64.0`
- `accept: */*`

Does this imply that the error is being generated by _cURL_ from within the container? Is this a _Docker_ health check? I couldn't find any guidance on this, so asking here. What am I doing wrong?

Edit: These email are being sent continuously, approximately five minutes apart.

**Reply by JC5** [reply]
Yep that is the healthcheck from the container. My apologies. This will keep happening until I rebuild the Docker container which takes at least an hour. Stay tuned.

**Reply by JC5** [reply]
I rebuilt `develop` and `5.7.10`, you'll have to pull them again to see the difference, I hope it works 😁

**Reply by JC5** [reply]
Apologies, the build is still running.

**Reply by viharm** [reply]
I have pulled a fresh copy of the image and launched again. Container has been up for 8 min, with no error emails. _Docker_ `ps` shows `(healthy)`

**Reply by JC5** [reply]
Sounds good! The build was suspiciously fast, so I clear the build-cache just to be sure. When you do that, it takes over an hour lol. If it works I'm not complaining but I'm doublechecking just in case.

**Reply by viharm** [reply]
OK, that's cool. Thanks for the prompt action :-)

Please let me know if it needs further testing.

I will consolidate all my configs to post a sample suite of configs as examples for the FF3-Docker, NGINX, Authelia, OpenLDAP combination - hopefully over the next few days. Should I post it here?

**Reply by JC5** [reply]
That would be amazing, thanks!

**Comment by madewithpat** [comment]
Hey @viharm - did you ever get sample configuration put anywhere?

**Reply by viharm** [reply]
Hi, sorry about the late response. I have got this working, but haven't yet sanitized my config to share yet. Will do it over the next few days.

**Reply by panoskpv** [reply]
@viharm Any luck with the sample configuration. I would like to add authelia to my firefly.

**Reply by viharm** [reply]
Hi @panoskpv (&James), sorry about the delay. Life keeps getting in the way.

I believe my new/improved circumstances will allow me to provide one this weekend.

**Comment by viharm** [comment, marked as answer]
I've finally finished my write up and it includes all _Firefly III_ settings needed to work with _Authelia_. Hope this helps.

https://cerebral.voyage/firefly-iii-on-docker-with-mariadb-redis-ldap/

**Reply by JC5** [reply]
Very cool!

**Comment by viharm** [comment]
cc @panoskpv , @madewithpat
